README.pod
=pod

=head1 NAME

Mojo::DB::Role::ResultsRoles - Apply roles to Mojo database results

=head1 SYNOPSIS

  use Mojo::Pg;
  my $pg = Mojo::Pg->new(...)->with_roles('Mojo::DB::Role::ResultsRoles');
  push @{$pg->results_roles}, 'Mojo::DB::Results::Role::Something';
  my $results = $pg->db->query(...);
  # $results does Mojo::DB::Results::Role::Something

=head1 DESCRIPTION

This role allows roles to be applied to the results objects returned by
L<Mojo::Pg> or similar database APIs. The manager object must create database
connections via a C<db> method, which must have a C<results_class> attribute
used to instantiate results objects.

=head1 ATTRIBUTES

L<Mojo::DB::Role::ResultsRoles> composes the following attributes.

=head2 results_roles

  my $roles = $manager->results_roles;
  $manager  = $manager->results_roles(\@roles);

Array reference of roles to compose into results objects. This only affects
database objects created by subsequent calls to the C<db> method.

=head1 BUGS

Report any issues on the public bugtracker.

=head1 AUTHOR

Dan Book <dbook@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2019 by Dan Book.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=head1 SEE ALSO

L<Mojo::Pg>, L<Mojo::mysql>, L<Mojo::SQLite>

=cut


Changes
{{$NEXT}}

v0.1.1    2021-06-15 23:53:56 EDT
  - Update IRC metadata to libera.chat

v0.1.0    2019-12-03 17:50:02 EST
  - First release


Build.PL
# This Build.PL for Mojo-DB-Role-ResultsRoles was generated by
# Dist::Zilla::Plugin::ModuleBuildTiny::Fallback 0.025
use strict;
use warnings;

my %configure_requires = (
    'Module::Build::Tiny' => '0.034',
);

my %errors = map {
    eval "require $_; $_->VERSION($configure_requires{$_}); 1";
    $_ => $@,
} keys %configure_requires;

if (!grep { $_ } values %errors)
{
    # This section for Mojo-DB-Role-ResultsRoles was generated by Dist::Zilla::Plugin::ModuleBuildTiny 0.015.
    use strict;
    use warnings;

    use 5.010001;
    # use Module::Build::Tiny 0.034;
    Module::Build::Tiny::Build_PL();
}
else
{
    if (not $ENV{PERL_MB_FALLBACK_SILENCE_WARNING})
    {
        warn <<'EOW'
*** WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING ***

If you're seeing this warning, your toolchain is really, really old* and you'll
almost certainly have problems installing CPAN modules from this century. But
never fear, dear user, for we have the technology to fix this!

If you're using CPAN.pm to install things, then you can upgrade it using:

    cpan CPAN

If you're using CPANPLUS to install things, then you can upgrade it using:

    cpanp CPANPLUS

If you're using cpanminus, you shouldn't be seeing this message in the first
place, so please file an issue on github.

This public service announcement was brought to you by the Perl Toolchain
Gang, the irc.perl.org #toolchain IRC channel, and the number 42.

----

* Alternatively, you are running this file manually, in which case you need
to learn to first fulfill all configure requires prerequisites listed in
META.yml or META.json -- or use a cpan client to install this distribution.

You can also silence this warning for future installations by setting the
PERL_MB_FALLBACK_SILENCE_WARNING environment variable, but please don't do
that until you fix your toolchain as described above.


Errors from configure prereqs:
EOW
        . do {
            require Data::Dumper; Data::Dumper->new([ \%errors ])->Indent(2)->Terse(1)->Sortkeys(1)->Dump;
        };

        sleep 10 if -t STDIN && (-t STDOUT || !(-f STDOUT || -c STDOUT));
    }


    # This section was automatically generated by Dist::Zilla::Plugin::ModuleBuild v6.012.
    use strict;
    use warnings;

    require Module::Build; Module::Build->VERSION(0.28);


    my %module_build_args = (
      "configure_requires" => {
        "Module::Build::Tiny" => "0.034"
      },
      "dist_abstract" => "Apply roles to Mojo database results",
      "dist_author" => [
        "Dan Book <dbook\@cpan.org>"
      ],
      "dist_name" => "Mojo-DB-Role-ResultsRoles",
      "dist_version" => "v0.1.1",
      "license" => "artistic_2",
      "module_name" => "Mojo::DB::Role::ResultsRoles",
      "recursive_test_files" => 1,
      "requires" => {
        "Class::Method::Modifiers" => 0,
        "Mojolicious" => "7.55",
        "Role::Tiny" => "2.000001",
        "perl" => "5.010001"
      },
      "test_requires" => {
        "File::Spec" => 0,
        "Module::Metadata" => 0,
        "Test::More" => "0.88",
        "Test::Needs" => 0
      }
    );


    my %fallback_build_requires = (
      "File::Spec" => 0,
      "Module::Metadata" => 0,
      "Test::More" => "0.88",
      "Test::Needs" => 0
    );


    unless ( eval { Module::Build->VERSION(0.4004) } ) {
      delete $module_build_args{test_requires};
      $module_build_args{build_requires} = \%fallback_build_requires;
    }

    my $build = Module::Build->new(%module_build_args);


    $build->create_build_script;
}


lib/Mojo/DB/Role/ResultsRoles.pm
package Mojo::DB::Role::ResultsRoles;

use Mojo::Base -role;

our $VERSION = 'v0.1.2';

requires 'db';

has results_roles => sub { [] };

around db => sub {
  my $orig = shift;
  my $self = shift;
  my $db = $self->$orig(@_);
  my $class = $db->results_class;
  my $roles = $self->results_roles;
  $db->results_class($class->with_roles(@$roles)) if @$roles;
  return $db;
};

1;

=head1 NAME

Mojo::DB::Role::ResultsRoles - Apply roles to Mojo database results

=head1 SYNOPSIS

  use Mojo::Pg;
  my $pg = Mojo::Pg->new(...)->with_roles('Mojo::DB::Role::ResultsRoles');
  push @{$pg->results_roles}, 'Mojo::DB::Results::Role::Something';
  my $results = $pg->db->query(...);
  # $results does Mojo::DB::Results::Role::Something

=head1 DESCRIPTION

This role allows roles to be applied to the results objects returned by
L<Mojo::Pg> or similar database APIs. The manager object must create database
connections via a C<db> method, which must have a C<results_class> attribute
used to instantiate results objects.

=head1 ATTRIBUTES

L<Mojo::DB::Role::ResultsRoles> composes the following attributes.

=head2 results_roles

  my $roles = $manager->results_roles;
  $manager  = $manager->results_roles(\@roles);

Array reference of roles to compose into results objects. This only affects
database objects created by subsequent calls to the C<db> method.

=head1 BUGS

Report any issues on the public bugtracker.

=head1 AUTHOR

Dan Book <dbook@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2019 by Dan Book.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=head1 SEE ALSO

L<Mojo::Pg>, L<Mojo::mysql>, L<Mojo::SQLite>


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.012.
use strict;
use warnings;

use 5.010001;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Apply roles to Mojo database results",
  "AUTHOR" => "Dan Book <dbook\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Mojo-DB-Role-ResultsRoles",
  "LICENSE" => "artistic_2",
  "MIN_PERL_VERSION" => "5.010001",
  "NAME" => "Mojo::DB::Role::ResultsRoles",
  "PREREQ_PM" => {
    "Class::Method::Modifiers" => 0,
    "Mojolicious" => "7.55",
    "Role::Tiny" => "2.000001"
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => "0.88"
  },
  "VERSION" => "v0.1.0",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Class::Method::Modifiers" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Mojolicious" => "7.55",
  "Role::Tiny" => "2.000001",
  "Test::More" => "0.88"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
